Descriptive error messages for missing template hooks errors

Vild opened the pull request:
This is my warmup project for GSoC that was suggested by my mentors. This PR implemented more descriptive error messages for missing template hooks. This replaces following the none-descriptive error message: `Error: undefined identifier __equals in module object` with:
> Error: object.__equals not found. The current runtime does not support equal checks on arrays, or the runtime is corrupt.

The error message is modeled after ObjectNotFound: https://github.com/dlang/dmd/blob/8e364d300adef422a36ac39b0f09dcfaed78312a/src/dmd/declaration.d#L202

Questions:
- Does verifyHookExist use the best way of checking for an identifier inside the `object` module?
- `ObjectNotFound` have the "or corrupt" part, should this error message also have it?
- Any improvements to make the error message more clear?

---

Output from this PR with a empty object.d and -checkaction=context:
```d
struct MyStruct { int a, b; }
MyStruct[] castToMyStruct(int[] arr) {
	return cast(MyStruct[])arr;	// Error: object.__ArrayCast not found. The current runtime does not support casting array of structs, or the runtime is corrupt.
}

extern(C) int main(int argc, char** argv) {
	int[] arrA, arrB;
	bool a = arrA[] == arrB[]; // Error: object.__equals not found. The current runtime does not support equal checks on arrays, or the runtime is corrupt.
	bool b = arrA < arrB; // Error: object.__cmp not found. The current runtime does not support comparing arrays, or the runtime is corrupt.

	{
		int x = 1; int y = 1;
		assert(x == y); // Error: object._d_assert_fail not found. The current runtime does not support generating assert messages, or the runtime is corrupt.
	}

	switch ("a") { // Error: object.__switch not found. The current runtime does not support switch cases on strings, or the runtime is corrupt.
	default:
		break;
	}

	final switch (0) { // Error: object.__switch_error not found. The current runtime does not support generating assert messages, or the runtime is corrupt.
	case 1:
		break;
	}

	return 0;
}

```

Vild committed:
Add verifyHookExist function

Signed-off-by: Dan Printzell <xwildn00bx@gmail.com>

wilzbach commented:
Nice! Could you add test cases for this, s.t. the new code is covered? Have a look at `test/run.d` and the readme in test for details

JinShil commented:
> std/json.d(702): Error: `object.__equals` not found. The current runtime does not support equal checks on arrays, or the runtime is corrupt.

That's strange.  Is it a false positive?

JinShil commented on src/dmd/expressionsem.d:
```diff
@@ -11746,3 +11772,25 @@ VarDeclaration makeThis2Argument(const ref Loc loc, Scope* sc, FuncDeclaration f
+ * Returns:
+ *      If the hook was found.
+ */
+extern (C++) bool verifyHookExist(const ref Loc loc, Scope* sc, Identifier id, const(char)* description, Identifier module_ = Id.object)
```
Favor `ref Scope sc` if possible.

JinShil replied:
Also, favor `string` instead of `const(char)*`.  We're trying to get rid of `char*` in DMD.

Vild replied:
> Favor `ref Scope sc` if possible.

Scope is stored as a pointer inside ExpressionSemanticVisitor.
I could change it to a ref, but then I need to dereference it each time `verifyHookExist` is called
Should I still do it?

> Also, favor `string` instead of `const(char)*`. We're trying to get rid of `char*` in DMD.

The `const(char)*` is sent all the way down to the C function `vsnprintf` via `OutBuffer`, so it need to be a `const(char)*`. I can't find a `tostringz` function so I could not got that way either.

JinShil replied:
> Should I still do it?

JinShil replied:
> I can't find a tostringz function so I could not got that way either.

Is it possible for you to use `f("%.*s", cast(int)s.length, s.ptr)` to workaround the current limitations of using `string` in DMD?  That technique is used in several other places in DMD currently.

jacob-carlborg replied:
Yes, `"%.*s"` will work.

jacob-carlborg replied:
> I can't find a tostringz function so I could not got that way either.

There's one here:

https://github.com/dlang/dmd/blob/df6418f0dd4cfeb79129a717f1a1037ccf4c90ae/src/dmd/utils.d#L207

JinShil commented:
> Does verifyHookExist use the best way of checking for an identifier inside the object module?

JinShil commented on src/dmd/expressionsem.d:
```diff
@@ -10653,10 +10673,16 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
-            Identifier id = Identifier.idPool("__equals");
             __equals = new DotIdExp(exp.loc, __equals, Id.object);
-            __equals = new DotIdExp(exp.loc, __equals, id);
+            __equals = new DotIdExp(exp.loc, __equals, Id.__equals);
```
It's best to put things like this in a separate PR.

Vild replied:
Will do

Vild commented:
> > std/json.d(702): Error: `object.__equals` not found. The current runtime does not support equal checks on arrays, or the runtime is corrupt.
> 
> That's strange. Is it a false positive?

The problem here is that when I search for `Id.object` it finds `std.json.JSONValue.object`.
I guess I need to find another way of referring directly to the `object` module.

wilzbach commented:
> I guess I need to find another way

Yes, but should be able to start with object and then use a dotTemplate instantiation for its members.

Vild committed:
Added testcase to verify `verifyHookExist` function

Signed-off-by: Dan Printzell <xwildn00bx@gmail.com>

Vild commented:
I've implemented fixed for everything so far:
- I had to start with `Id.empty` and then go to `object`, to correctly find `object.d`
- I removed the small refactoring that JinShin said
- verifyHookExist now take `Scope` as a ref, and `description` as a string.
  - I used `%.*s` to print the string, as suggested.
- I added a small failure test-case that verifies that all the error messages are triggered and outputted correctly.

jacob-carlborg commented on src/dmd/expressionsem.d:
```diff
@@ -11748,3 +11764,28 @@ VarDeclaration makeThis2Argument(const ref Loc loc, Scope* sc, FuncDeclaration f
+bool verifyHookExist(const ref Loc loc, ref Scope sc, Identifier id, string description, Identifier module_ = Id.object)
+{
+    Dsymbol rootSymbol = sc.search(loc, Id.empty, null);
+    Dsymbol moduleSymbol = rootSymbol.search(loc, module_);
```
* This could be moved into the if-statement
* `auto` could be used

Vild replied:
Better like this?
```d
auto rootSymbol = sc.search(loc, Id.empty, null);
if (auto moduleSymbol = rootSymbol.search(loc, module_))
    if (moduleSymbol.search(loc, id))
        return true;
```

jacob-carlborg replied:
Possibly, I'm not sure.

jacob-carlborg commented on src/dmd/expressionsem.d:
```diff
@@ -11748,3 +11764,28 @@ VarDeclaration makeThis2Argument(const ref Loc loc, Scope* sc, FuncDeclaration f
+    if (moduleSymbol)
+    {
+      Dsymbol idSymbol = moduleSymbol.search(loc, id);
+      if (idSymbol)
```
Indentation looks wrong.

jacob-carlborg commented on test/fail_compilation/verifyhookexist.d:
```diff
@@ -0,0 +1,47 @@
+---
+*/
+
+#line 100
```
Please avoid using these `#line` tokens. There's a command to automatically update the error messages: https://github.com/dlang/dmd/blob/master/test/README.md#automatically-update-the-test_output-segments

Vild committed:
Use verifyHookExist to verify that template runtime hooks exist

Signed-off-by: Dan Printzell <xwildn00bx@gmail.com>

Basile-z commented on src/dmd/expressionsem.d:
```diff
@@ -5718,6 +5718,10 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
                 tok == TOK.in_ ||
                 isEqualsCallExpression)
             {
+                // Verify that the hook `object._d_assert_fail` exist
```
The comment is useless because the function self documents enough

Basile-z replied:
ditto for the other similar code comments placed before a call to the check.

Basile-z commented on src/dmd/expressionsem.d:
```diff
@@ -11748,3 +11764,24 @@ VarDeclaration makeThis2Argument(const ref Loc loc, Scope* sc, FuncDeclaration f
+ *      description = what the hook does
+ *      module_ = what module the hook is located in
+ * Returns:
+ *      If the hook was found.
```
``` 
`true` if ... otherwise `false` 
```

or maybe

```
a `bool` indicating if the hook is present
```

Basile-z commented on src/dmd/expressionsem.d:
```diff
@@ -11748,3 +11764,24 @@ VarDeclaration makeThis2Argument(const ref Loc loc, Scope* sc, FuncDeclaration f
 }
+
+/*******************************
+ * Make sure that the runtime hook exists.
```
`a runtime hook` rather than `the`

JinShil replied:
I disagree with that.  "Make sure a runtime hook exists" means you are checking for the existence of at least 1 runtime hook.  "Make sure the runtime hook exists" means you are checking for the existence of a specific runtime hook.

Basile-z replied:
I meant that the function is not specilized for a particular hook, i.e "the hook" it checks "a hook" specified with a parameter.

Vild replied:
I'm changing it to `Make sure that the runtime hook $(PARAM id) exists.` to specify that it looks for the hook `id`.

jacob-carlborg replied:
Or `Make sure that the given runtime hook exists`. Now it can be confusing that it looks explicitly for a hook named "id".

Vild commented:
> Also in the error message is mentioned a possible corruption of the runtime. Can this really happen ?

I based it on the error that happens if TypeInfo is missing: `TypeInfo not found. object.d may be incorrectly installed or corrupt.`.

Another reason for keeping this part is in case the error get triggered, and the user does not really know what a runtime is, it is probably better to say that the runtime could be corrupted instead of just saying that the current runtime does not support feature *X*.

I guess it could be changed to something like `The current runtime does not support [...], or runtime version mismatch.`, but I think this is less clear than the corrupt part.

Basile-z commented:
lgtm

thewilsonator commented:
@Vild this good to go?

Vild commented:
> @Vild this good to go?

Yup, I'm happy with the PR.
